python/python/bystro/ancestry/tests/test_ancestry_types.py

```diff
@@ -0,0 +1,276 @@
+"""Test ancestry_types.py."""
+import pytest
+from attrs.exceptions import FrozenInstanceError
+
+from ..ancestry_types import (
+    AncestryResponse,
+    AncestryResult,
+    AncestrySubmission,
+    AttrValidationError,
+    PopulationVector,
+    ProbabilityInterval,
+    SuperpopVector,
+)
+
+POPULATIONS = [
+    "ACB",
+    "ASW",
+    "BEB",
+    "CDX",
+    "CEU",
+    "CHB",
+    "CHS",
+    "CLM",
+    "ESN",
+    "FIN",
+    "GBR",
+    "GIH",
+    "GWD",
+    "IBS",
+    "ITU",
+    "JPT",
+    "KHV",
+    "LWK",
+    "MAG",
+    "MSL",
+    "MXL",
+    "PEL",
+    "PJL",
+    "PUR",
+    "STU",
+    "TSI",
+    "YRI",
+]
+
+SUPERPOPS = [
+    "AFR",
+    "AMR",
+    "EAS",
+    "EUR",
+    "SAS",
+]
+
+
+def test_AncestrySubmission():
+    AncestrySubmission("foo.vcf")
+    AncestrySubmission("foo.vcf.gz")
+    AncestrySubmission("foo.vcf.lz4")
+    AncestrySubmission("foo.vcf.zstd")
+    AncestrySubmission("foo.vcf.bzip2")
+
+
+def test_AncestrySubmission_bad_filepaths():
+    with pytest.raises(AttrValidationError):
+        AncestrySubmission(3)
+    with pytest.raises(AttrValidationError):
+        AncestrySubmission("foo.txt")
+    with pytest.raises(AttrValidationError):
+        AncestrySubmission("foo.gz")
+    with pytest.raises(AttrValidationError):
+        AncestrySubmission("foo.vcf.docx")
+    with pytest.raises(AttrValidationError):
+        AncestrySubmission("foo.txt.gz")
+
+
+def test_AncestrySubmission_frozen():
+    ancestry_submission = AncestrySubmission("foo.vcf")
+    with pytest.raises(FrozenInstanceError):
+        ancestry_submission.vcf_path = "bar.vcf"
+
+
+def test_ProbabilityInterval() -> None:
+    """Ensure we can instantiate, validate ProbabilityInterval correctly."""
+    ProbabilityInterval(lower_bound=0.1, upper_bound=0.9)
+
+    with pytest.raises(AttrValidationError):
+        ProbabilityInterval(lower_bound=0.1, upper_bound=1.1)
+
+    with pytest.raises(AttrValidationError):
+        ProbabilityInterval(lower_bound=-2, upper_bound=1.1)
+
+    with pytest.raises(AttrValidationError):
+        ProbabilityInterval(lower_bound=1, upper_bound=0)
+
+    prob_int = ProbabilityInterval(lower_bound=0.0, upper_bound=1.0)
+    with pytest.raises(FrozenInstanceError):
+        prob_int.lower_bound = 0.5  # type: ignore [misc]
+
+
+prob_int = ProbabilityInterval(lower_bound=0.0, upper_bound=1.0)
```

Comment:
prob_int is defined right above as the same thing - ignore this if it's included here as well for a separate reason but otherwise one of them could be removed.